### PR TITLE
Light client state iteration proof.

### PR DESCRIPTION
**Pull request by cheme**
This PR implements state iteration for light client.
It should also provide an implementation for 'state_getKeysPaged' rpc for light clients.

There is a lot of boilerplate code but most logic is in prove and check function in state-machine.

Also depends on https://github.com/paritytech/trie/pull/127 (could be implemented without but would fetch and additional node in all proofs).

### CODE REVIEW

**Comment by cheme** [comment]
CI is blocked behind https://github.com/paritytech/parity-scale-codec/pull/253 but tests are runing locally.

**Comment by cheme** [comment]
Also related to https://github.com/paritytech/smoldot/pull/576 CC\ @tomaka

**Comment by cheme** [comment]
I did put the PR under 'D5-nice-to-have-audit', because the changes to state-machine, even if pretty trivial are exposing an api which breaks on error while the common api did skip error, so even if it looks rather trivial, this is related to some existing assumption of state-machine assertion.

**Comment by cheme** [comment]
This PR should not change any existing behavior for runtime or externalities function. In this sense runtime audit should just be checking that point, it is actually pretty trivial so I switch back to 'D1-trivial'.

**Comment by arkpar** [comment]
Limiting response by key count rather than byte size may cause trouble. You never now if given 1000 keys are over the hardcoded packet limit, which will result in error. If this is used to reconstruct the whole state, the syncing algorithm would need to account for such an error and reduce requested count.

File: client/light/src/backend.rs [hunk 1]
```diff
@@ -461,6 +461,21 @@ impl<H: Hasher> StateBackend<H> for GenesisOrUnavailableState<H>
 		}
 	}
 
+	fn apply_to_key_values_while<A: FnMut(&[u8], &[u8]) -> bool>(
```
Comment (arkpar) [inline client/light/src/backend.rs:464]: Could be `FnMut(Vec<u8>, Vec<u8>)` to avoid copying data

File: client/api/src/light.rs [hunk 2]
```diff
@@ -90,6 +90,29 @@ pub struct RemoteReadChildRequest<Header: HeaderT> {
...
+	pub header: Header,
+	/// Storage key of child trie to query, `None` for top trie.
+	pub child_trie_key: Option<PrefixedStorageKey>,
+	/// Limit range of query to a given prefix.
+	pub prefix: Option<Vec<u8>>,
+	/// Maximum number of key value to query.
+	pub count: u32,
+	/// Maximum total size of value to query.
+	/// Acts as a threshold.
+	pub value_size: u32,
```
Comment (arkpar) [inline client/api/src/light.rs:108]: I'd suggest this to be called `max_value_size` or `value_size_limit` here and everywhere.

**Review by arkpar — APPROVED** [review]

**Pull request closed without merging**